### ISSUE TITLE
Weird variable scope

### ISSUE THREAD

**Issue by Kjarrigan** [Flipez/rocket-lang#13]
The title is a bit weird because I'm not sure how to describe it properly in brief BUT for the current AoC I tried RL once again and had some issues with the missing string functions so I tried building some myself, but due to how the variables work it's currently not possible. See my sample:

```javascript
"// Last Updated: 2021.12.02"
"// RocketLang Version: 0.9.7"
"------------------------------------"

let x = 0;
let depth = 0;

let each = fn(enumerable, yield, idx) {
  if (idx == len(enumerable)) {
    return len(enumerable);
  }
  yield(enumerable[idx]);
  each(enumerable, yield, idx + 1);
};

"Does not work because you have to do push with let a = push(a, b)"
"which makes a a 'local' variable and thus breaks everything -.-"
let split = fn(string, separator) {
  let list = [];
  let current_word = "";

  each(string, fn(char) {
    puts(char == separator);
    if (char == separator) {
      puts "Split found!"
      let list = push(list, current_word);
      let current_word = "";
    } else {
      puts(current_word);
      puts("Another letter to the word: " + char);
      let current_word = current_word + char;
      puts(current_word);
    }
  }, 0);
  let list = push(list, current_word);

  return list;
};

split("Hello World", " ");

"Simplified sample of split. super sad that this does not work :-("
let char_array = [];
let counter = 0;
each("Hello World", fn(c) {
  let char_array = push(char_array, c);
  let counter = counter + 1;
}, 0);
puts(char_array);
puts(counter);
```

In theory it should "fill" the outer variables with characters and strings because they are defined outside of the inner block but I can't do `current_word = current_word + "Hello"` because I have to write a `let` before which then makes it a local variable thus the house of cards crumble.

Not sure if this is solvable but a workaround might be to make push change the array itself rather than returning a new one, but I think you tried this already when I requested the push/pop functions....

**Comment by Kjarrigan** [comment]
if this could be solved a lot of (helper) functions could be implemented like:
```javascript
"Mississippi".count("i") # => 4
"Hello World".split(" ") # => ["Hello", "World"]
"Hello World".tr("e", "a") # => "Hallo World"
```

**Comment by Flipez** [comment]
All the issues mentioned should be fixed in version 0.10.0 as it introduces a lot new features such as
  - in-place operations
  - foreach loop
  - let-free variable usage

Feel free to test with the next branch until release.
